### PR TITLE
Feat nrf time rtc support

### PR DESCRIPTION
**Pull request by hoihu**
This is a cleanup version of the inputs collected in https://github.com/micropython/micropython/pull/6171
(I didn't want to force push the update to that PR because it contains implementation details
via ticker module that might be useful for others)

Basically it adds support for time.ticks_ms / time.ticks_us via RTC1, configured in 30usec tick mode. It also 
adds the time.ticks_add and time.ticks_diff helper functions.

For this to work properly the RTC1's overflow irq is handled and a msec / usec ticks counter appropiately 
generated (see also commit comment for this relatively complex matter). Thanks for @dpgeorge for 
basically providing the solution there!

The PR also contains support for MICROPY_EVENT_POLL_HOOK. The wakeup from WFI is done via RTC1's 
CCR0, configured in app. 1msec intervals. Basically that makes the nrf port ready for asyncio.

This feature can be enabled using MICROPY_PY_TIME_TICKS. If disabled, the system uses the legacy 
ticks and sleep methods.

### CODE REVIEW

File: ports/nrf/modules/machine/rtcounter.c [hunk 1]
```diff
@@ -30,6 +30,7 @@
 #include "rtcounter.h"
 #include "nrfx_rtc.h"
 #include "nrf_clock.h"
+#include "py/mperrno.h"
```
Comment (dpgeorge) [inline ports/nrf/modules/machine/rtcounter.c:33]: I guess this addition is not needed

File: ports/nrf/mphalport.c [hunk 2]
```diff
@@ -35,6 +35,113 @@
...
+};
+
+STATIC void rtc_irq_time(nrfx_rtc_int_type_t event) {
+    // irq handler for overflow
+    if (event == NRFX_RTC_INT_OVERFLOW) {
+        rtc_overflows += 1;
+    }
+    // irq handler for wakeup from WFI (~1msec)
+    if (event == NRFX_RTC_INT_COMPARE0) {
+        nrfx_rtc_cc_set(&rtc1, 0, nrfx_rtc_counter_get(&rtc1) + 33, true);
```
Comment (dpgeorge) [inline ports/nrf/mphalport.c:70]: the "33" should probably be a #define constant

File: ports/nrf/mphalport.c [hunk 3]
```diff
@@ -35,6 +35,113 @@
...
+
+void rtc1_init_time_ticks(void) {
+    // Start the low-frequency clock (if it hasn't been started already)
+    if (!nrf_clock_lf_is_running(NRF_CLOCK)) {
+        nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_LFCLKSTART);
+    }
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
+    nrfx_rtc_overflow_enable(&rtc1, true);
+    // generate a ~1msec irq to leave _WFI
+    nrfx_rtc_cc_set(&rtc1, 0, 30, true);
```
Comment (dpgeorge) [inline ports/nrf/mphalport.c:82]: the "30" should be the same as the "33" above, right?  in which case use the same #define constant

File: ports/nrf/mphalport.c [hunk 4]
```diff
@@ -70,7 +177,11 @@ int mp_hal_stdin_rx_chr(void) {
         if (MP_STATE_PORT(board_stdio_uart) != NULL && uart_rx_any(MP_STATE_PORT(board_stdio_uart))) {
             return uart_rx_char(MP_STATE_PORT(board_stdio_uart));
         }
+        #if MICROPY_PY_TIME_TICKS
+        MICROPY_EVENT_POLL_HOOK
+        #else
```
Comment (dpgeorge) [inline ports/nrf/mphalport.c:182]: could it just use `MICROPY_EVENT_POLL_HOOK` unconditionally (ie in both cases here)?

**Comment by dpgeorge** [comment]
Looks really good!

File: ports/nrf/mphalport.c [hunk 5]
```diff
@@ -35,6 +35,113 @@
...
+        nrfx_rtc_cc_set(&rtc1, 0, nrfx_rtc_counter_get(&rtc1) + 33, true);
+    }
+}
+
+void rtc1_init_time_ticks(void) {
+    // Start the low-frequency clock (if it hasn't been started already)
+    if (!nrf_clock_lf_is_running(NRF_CLOCK)) {
+        nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_LFCLKSTART);
+    }
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
```
Comment (glennrub) [inline ports/nrf/mphalport.c:79]: Should we add a `nrfx_rtc_counter_clear()` here? Right now, it remembers the old counter value across `soft reset`. Or is it more interesting to have the count from power on?

Comment (dpgeorge) [reply ports/nrf/mphalport.c:79]: I think just leave it, to keep it simple.  Ticks should anyway not be used as an absolute measure of time, just for relative differences (that's how it works in most other ports, it's only reset on power-up).

Comment (glennrub) [reply ports/nrf/mphalport.c:79]: I agree. I also noticed the documentation after writing the question which says: "... counter with an arbitrary reference point ...".

**Comment by glennrub** [comment]
From my side, it looks good as well.
I've tested the patch a bit, including WFI, overflow and my obscure "IRQ in IRQ" scenario. I all seems to work very well. I've also test integrated it into my nrf91 sockets, and you have solved my need for timer. Thank you @hoihu !

File: ports/nrf/modules/machine/rtcounter.c (hunk 1 above)
Comment (hoihu) [reply ports/nrf/modules/machine/rtcounter.c:33]: ok

File: ports/nrf/mphalport.c (hunk 2 above)
Comment (hoihu) [reply ports/nrf/mphalport.c:70]: yes

File: ports/nrf/mphalport.c (hunk 3 above)
Comment (hoihu) [reply ports/nrf/mphalport.c:82]: that would be the first CC irq to schedule, so can actually be different? 

But I haven't looked at it closely. It seems that nrfx_rtc_init does not clear the counter state (as I have assumed). At least I didn't find any reference to it.

So I believe we should reset the counter using `nrfx_rtc_counter_clear`, as pointed below from @glennrub  of use a similar logic like `nrfx_rtc_counter_get(&rtc1) + xxx` ? My preference would be the second one.

File: ports/nrf/mphalport.c (hunk 4 above)
Comment (hoihu) [reply ports/nrf/mphalport.c:182]: Yes I was thinking about that too.. (since there was already a WFI statement). But I wanted it to behave like the old, legacy solution if `MICROPY_PY_TIME_TICKS` was disabled. But I can easily enable it always, no problem.

File: ports/nrf/mphalport.c (hunk 5 above)
Comment (hoihu) [reply ports/nrf/mphalport.c:79]: > Should we add a nrfx_rtc_counter_clear() here?

see comment above

**Comment by hoihu** [comment]
> From my side, it looks good as well.
I've tested the patch a bit, including WFI, overflow and my obscure "IRQ in IRQ" scenario. I all seems to work very well. I've also test integrated it into my nrf91 sockets, and you have solved my need for timer. Thank you @hoihu !

Great, thanks!

File: ports/nrf/mphalport.c (hunk 3 above)
Comment (dpgeorge) [reply ports/nrf/mphalport.c:82]: > My preference would be the second one.

Sounds good.

File: ports/nrf/mphalport.c (hunk 4 above)
Comment (dpgeorge) [reply ports/nrf/mphalport.c:182]: aah, since `MICROPY_ENABLE_SCHEDULER` is disabled on nrf it won't make a difference, `mp_handle_pending()` won't do anything except check for pending exceptions, but at the REPL (where this function is used) ctrl-C (the only possible pending exception) is disabled

Comment (hoihu) [reply ports/nrf/mphalport.c:182]: ok

**Comment by hoihu** [comment]
I force pushed the review feedbacks, changes are:
- `MICROPY_EVENT_POLL_HOOK` is enabled unconditionally in `mp_hal_stdin_rx_chr`
- The first RTC's CC irq is now properly scheduled (I used a macro there too). The RTC counter is soft-reset proof (not sure if that really matters, but might be nice on the REPL).
- Deleted unused #include

From my point of view ready to go in.

**Comment by dpgeorge** [comment]
Thanks for updating, and for the effort putting this together.  Now merged in 15574cd665a8e66157aa97447e3869513da81d45 and 59ed3bdd9fa8dc6e0677cf8d4478a5a32fee4500 (with minor edits to reorder some bits, and remove `MICROPY_PY_TIME_TICKS` guard from `MICROPY_EVENT_POLL_HOOK` definition).
